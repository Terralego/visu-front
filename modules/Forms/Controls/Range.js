import React from 'react';
import PropTypes from 'prop-types';
import { RangeSlider } from '@blueprintjs/core';

import './index.scss';

export const Range = ({
  label,
  values,
  value,
  onChange,
  ...props
}) => {
  const val = value.length ? value : values;
  if (val.length !== 2 || (+val[0] > +val[1])) {
    throw new Error('Range control: There must be two values and the first must be less than the second');
  }
  return (
    <div className="control-container">
      <p className="control-label">{label}</p>
      <RangeSlider
        value={val}
        onChange={onChange}
        {...props}
      />
    </div>
  );
};

Range.propTypes = {
  label: PropTypes.string,
  value: PropTypes.arrayOf(PropTypes.number),
  values: PropTypes.arrayOf(PropTypes.number),
  onChange: PropTypes.func,
  min: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,
};

Range.defaultProps = {
  label: '',
  value: [],
  values: [],
  onChange () {},
};

export default Range;
