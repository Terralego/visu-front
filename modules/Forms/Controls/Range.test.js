import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Range from './Range';

jest.mock('@blueprintjs/core', () => ({
  RangeSlider: () => <p>Range</p>,
}));

it('should render correctly', () => {
  const tree = renderer.create((
    <Range
      label="My range"
      values={[1, 4]}
      onChange={() => null}
      min={0}
      max={10}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should mount & update correctly', () => {
  const onChange = jest.fn();
  const wrapper = shallow((
    <Range
      label="Change Values of my range"
      onChange={onChange}
      values={[10, 30]}
      min={0}
      max={100}
    />
  ));
  wrapper.find('RangeSlider').props().onChange([20, 40]);
  expect(onChange).toHaveBeenCalled();
});

it('should throw an error when values are incorrect', () => {
  const ERROR_MESSAGE =  'Range control: There must be two values and the first must be less than the second';
  expect(() => Range({ value: [10] })).toThrow(ERROR_MESSAGE);
  expect(() => Range({ value: [30, 10] })).toThrow(ERROR_MESSAGE);
});
