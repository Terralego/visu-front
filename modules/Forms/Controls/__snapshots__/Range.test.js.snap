// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="control-container"
>
  <p
    className="control-label"
  >
    My range
  </p>
  <p>
    Range
  </p>
</div>
`;
